crates/sui-core/src/transaction_manager.rs

```diff
@@ -93,7 +95,134 @@ impl LockQueue {
     }
 }
 
-#[derive(Default)]
+struct CacheInner {
+    versioned_cache: LruCache<ObjectID, SequenceNumber>,
+    // we cache packages separately, because they are more expensive to look up in the db, so we
+    // don't want to evict packages in favor of mutable objects.
+    unversioned_cache: LruCache<ObjectID, ()>,
+
+    metrics: Arc<AuthorityMetrics>,
+}
+
+impl CacheInner {
+    fn bounded(size: usize, metrics: Arc<AuthorityMetrics>) -> Self {
+        Self {
+            versioned_cache: LruCache::new(size.try_into().unwrap()),
+            unversioned_cache: LruCache::new(size.try_into().unwrap()),
+            metrics,
+        }
+    }
+
+    fn unbounded(metrics: Arc<AuthorityMetrics>) -> Self {
+        Self {
+            versioned_cache: LruCache::unbounded(),
+            unversioned_cache: LruCache::unbounded(),
+            metrics,
+        }
+    }
+}
+
+impl CacheInner {
+    fn insert(&mut self, object: &InputKey) {
+        if let Some(version) = object.1 {
+            if let Some((previous_id, previous_version)) =
+                self.versioned_cache.push(object.0, version)
+            {
+                if previous_id == object.0
+                    && previous_version > object.1.unwrap_or(SequenceNumber::MIN)
+                {
+                    // do not allow highest known version to decrease
+                    self.versioned_cache.put(object.0, previous_version);
+                } else {
+                    self.metrics
+                        .transaction_manager_object_cache_evictions
+                        .inc();
+                }
+            }
+        } else if let Some((previous_id, _)) = self.unversioned_cache.push(object.0, ()) {
+            if previous_id != object.0 {
+                self.metrics
+                    .transaction_manager_package_cache_evictions
+                    .inc();
+            }
+        }
+    }
+
+    // Returns Some(true/false) for a definitive result. Returns None if the caller must defer to
+    // the db.
+    fn is_object_available(&mut self, object: &InputKey) -> Option<bool> {
+        if let Some(version) = object.1 {
+            if let Some(current) = self.versioned_cache.get(&object.0) {
+                self.metrics.transaction_manager_object_cache_hits.inc();
+                Some(*current >= version)
+            } else {
+                self.metrics.transaction_manager_object_cache_misses.inc();
+                None
+            }
+        } else {
+            self.unversioned_cache
+                .get(&object.0)
+                .tap_some(|_| self.metrics.transaction_manager_package_cache_hits.inc())
+                .tap_none(|| self.metrics.transaction_manager_package_cache_misses.inc())
+                .map(|_| true)
+        }
+    }
+}
+
+struct AvailableObjectsCache {
+    bounded_cache: CacheInner,
+    unbounded_cache: Option<CacheInner>,
```

Comment:
I'm thinking if it is possible and more straightforward to keep a single unbounded cache, and evict explicitly when unbounded_cache_enabled is 0 or under another condition. This would make reading and writing to the cache more efficient.

Reply:
very good idea